runtime/src/network/iouring.rs

```diff
@@ -282,69 +298,144 @@ impl crate::Sink for Sink {
+            {
+                // Channel closed - io_uring thread died, buffer is lost
+                return (StableBuf::default(), Err(crate::Error::RecvFailed));
+            }
+
+            let Ok((result, buf)) = rx.await else {
+                // Channel closed - io_uring thread died, buffer is lost
+                return (StableBuf::default(), Err(crate::Error::RecvFailed));
+            };
+            buffer = buf.unwrap();
 
-            // Wait for the operation to complete
-            let (result, got_buf) = rx.await.map_err(|_| crate::Error::RecvFailed)?;
-            buf = got_buf.unwrap();
             if should_retry(result) {
                 continue;
             }
 
-            // Non-positive result indicates an error or EOF.
             if result <= 0 {
-                return Err(crate::Error::RecvFailed);
+                let err = if result == -libc::ETIMEDOUT {
+                    crate::Error::Timeout
+                } else {
+                    crate::Error::RecvFailed
+                };
+                return (buffer, Err(err));
+            }
+
+            return (buffer, Ok(result as usize));
+        }
+    }
+
+    /// Fills the internal buffer by reading from the socket via io_uring.
+    async fn fill_buffer(&mut self) -> Result<usize, crate::Error> {
+        self.buffer_pos = 0;
+        self.buffer_len = 0;
+
+        let buffer: StableBuf = std::mem::take(&mut self.buffer).into();
+        let len = buffer.len();
+
+        // If the buffer is lost due to a channel error, we don't restore it.
+        // Channel errors mean the io_uring thread died, so the stream is unusable anyway.
+        let (buffer, result) = self.submit_recv(buffer, 0, len).await;
+        self.buffer = buffer.into();
+        self.buffer_len = result?;
+        Ok(self.buffer_len)
+    }
+}
+
+impl crate::Stream for Stream {
+    async fn recv(&mut self, buf: impl Into<StableBuf> + Send) -> Result<StableBuf, crate::Error> {
+        let mut buf = buf.into();
+        let mut bytes_received = 0;
+        let buf_len = buf.len();
+
+        while bytes_received < buf_len {
+            // First drain any buffered data
+            let buffered = self.buffer_len - self.buffer_pos;
+            if buffered > 0 {
+                let to_copy = std::cmp::min(buffered, buf_len - bytes_received);
+                buf.as_mut()[bytes_received..bytes_received + to_copy]
+                    .copy_from_slice(&self.buffer[self.buffer_pos..self.buffer_pos + to_copy]);
+                self.buffer_pos += to_copy;
+                bytes_received += to_copy;
+                continue;
+            }
+
+            let remaining = buf_len - bytes_received;
+
+            // Buffer is empty, read from socket
+            // For large reads or unbuffered mode, read directly into caller's buffer
+            let buffer_capacity = self.buffer.capacity();
+            if buffer_capacity == 0 || remaining >= buffer_capacity {
```

Comment:
nit: add a small comment indicating this is just handling the empty buffer case?